sdk/identity/azure-identity/azure/identity/_credentials/token_exchange.py

```diff
@@ -0,0 +1,68 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import os
+import time
+from typing import TYPE_CHECKING
+
+from .client_assertion import ClientAssertionCredential
+from .. import CredentialUnavailableError
+from .._constants import EnvironmentVariables
+
+if TYPE_CHECKING:
+    # pylint:disable=unused-import,ungrouped-imports
+    from typing import Any
+    from azure.core.credentials import AccessToken
+
+
+class TokenFileMixin(object):
+    def __init__(self):
+        # type: () -> None
+        super(TokenFileMixin, self).__init__()
+        self._jwt = ""
+        self._last_read_time = 0
+        self._token_file_path = os.environ[EnvironmentVariables.TOKEN_FILE_PATH]
+
+    def get_service_account_token(self):
+        # type: () -> str
+        now = int(time.time())
+        if now - self._last_read_time > 300:
+            with open(self._token_file_path) as f:
+                self._jwt = f.read()
+            self._last_read_time = now
+        return self._jwt
+
+
+class TokenExchangeCredential(TokenFileMixin):
+    def __init__(self, **kwargs):
+        # type: (**Any) -> None
+        super(TokenExchangeCredential, self).__init__()
+        self._available = all(os.environ.get(var) for var in EnvironmentVariables.TOKEN_EXCHANGE_VARS)
+        if self._available:
```

Comment:
Summarizing offline conversation: it's moot because ManagedIdentityCredential is the public API of this credential and already logs the interesting bit (what platform it believes it's running on) and will only instantiate this credential when its required environment variables are set. TokenExchangeCredential can therefore assume its prerequisites are met, and stop trying to raise a useful error message when that isn't the case. Commit to that effect forthcoming.